subprojects/core/src/main/java/org/gradle/api/internal/file/collections/ImmutableFileCollection.java

```diff
@@ -0,0 +1,110 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.file.collections;
+
+import com.google.common.collect.ImmutableSet;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.file.AbstractFileCollection;
+import org.gradle.api.internal.file.IdentityFileResolver;
+import org.gradle.internal.file.PathToFileResolver;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+public class ImmutableFileCollection extends AbstractFileCollection {
+    private static final String DEFAULT_DISPLAY_NAME = "immutable file collection";
+    private static final EmptyImmutableFileCollection EMPTY = new EmptyImmutableFileCollection();
+
+    private final Set<Object> files;
+    private final String displayName;
+    private final PathToFileResolver resolver;
+
+    public static ImmutableFileCollection of(Object... files) {
+        return of(Arrays.asList(files));
+    }
+
+    public static ImmutableFileCollection of(Collection<?> files) {
+        return usingResolver(new IdentityFileResolver(), files);
+    }
+
+    public static ImmutableFileCollection usingResolver(PathToFileResolver fileResolver, Object[] files) {
+        return usingResolver(fileResolver, Arrays.asList(files));
+    }
+
+    public static ImmutableFileCollection usingResolver(PathToFileResolver fileResolver, Collection<?> files) {
+        if (files.isEmpty()) {
+            return EMPTY;
+        }
+        return new ImmutableFileCollection(DEFAULT_DISPLAY_NAME, fileResolver, files);
+    }
+
+    private ImmutableFileCollection(String displayName, PathToFileResolver fileResolver, Collection<?> files) {
+        this.displayName = displayName;
+        this.resolver = fileResolver;
+        ImmutableSet.Builder<Object> filesBuilder = ImmutableSet.builder();
+        if (files != null) {
+            filesBuilder.addAll(files);
+        }
+        this.files = filesBuilder.build();
+    }
+
+    @Override
+    public String getDisplayName() {
+        return displayName;
+    }
+
+    @Override
+    public Set<File> getFiles() {
+        DefaultFileCollectionResolveContext context = new DefaultFileCollectionResolveContext(new IdentityFileResolver());
+        FileCollectionResolveContext nested = context.push(resolver);
+        nested.add(files);
+
+        List<Set<File>> fileSets = new LinkedList<Set<File>>();
+        int fileCount = 0;
+        for (FileCollection collection : context.resolveAsFileCollections()) {
+            Set<File> files = collection.getFiles();
+            fileCount += files.size();
+            fileSets.add(files);
+        }
+        Set<File> allFiles = new LinkedHashSet<File>(fileCount);
+        for (Set<File> fileSet : fileSets) {
+            allFiles.addAll(fileSet);
+        }
```

Comment:
I wonder if this double-iteration is worth it.

Reply:
(again) originally came from `CompositeFileCollection`, where a016dbc85c0c14bbe7d1ca7b6fce596f019a5d7c switched to doing this double-iteration.

Reply:
Let's take a look at that guy separately.

Reply:
In any case, we should extract shared code into some shared location.